_posts/2018-08-14-grpc-with-protobuf.md

```diff
@@ -0,0 +1,225 @@
+      }
+
+      @Override
+      public T parse(InputStream stream) {
+        return gson.fromJson(new InputStreamReader(stream, StandardCharsets.UTF_8), clz);
+      }
+    };
+  }
+```
+
+Given a `Class` object for a some request or response, this function will produce a marshaller.  Using the marshallers, we can compose a full `MethodDescriptor` for each of the four CRUD methods.  Here is an example of the Method descriptor for _Create_:
+
+```java
+  static final MethodDescriptor<CreateRequest, CreateResponse> CREATE_METHOD =
+      MethodDescriptor.newBuilder(
+          marshallerFor(CreateRequest.class),
+          marshallerFor(CreateResponse.class))
+          .setFullMethodName(
+              MethodDescriptor.generateFullMethodName(SERVICE_NAME, "Create"))
+          .setType(MethodType.UNARY)
+          .build();
+```
+
+Note that if we were using Protobuf, we would use the existing Protobuf marshaller, and the method descriptors would be generated automatically.
+
+## Sending RPCs
+
+We need to update our [`KvClient`](https://github.com/carl-mastrangelo/kvstore/blob/b225d28c7c2f3c356b0f3753384b3329f2ab5911/src/main/java/io/grpc/examples/KvClient.java#L98), the gRPC client used in the previous post, to use our MethodDescriptors.  Additionally, since we won't be using any Protobuf types, the code needs to use `ByteBuffer` rather than `ByteString`.  That said, we can still use the `grpc-stub` package on Maven to issue the RPC.  Using the _Create_ method again as an example, here's how to make an RPC:
+
+```java
+    ByteBuffer key = createRandomKey();
+    ClientCall<CreateRequest, CreateResponse> call =
+        chan.newCall(KvGson.CREATE_METHOD, CallOptions.DEFAULT);
+    KvGson.CreateRequest req = new KvGson.CreateRequest();
+    req.key = key.array();
+    req.value = randomBytes(MEAN_VALUE_SIZE).array();
+
+    ListenableFuture<CreateResponse> res = ClientCalls.futureUnaryCall(call, req);
+    // ...
+```
+
+As you can see, we create a new `ClientCall` object from the `MethodDescriptor`, create the request, and then send it using `ClientCalls.futureUnaryCall` in the stub library.  gRPC takes care of the rest for us.  You can also make blocking stubs or async stubs instead of future stubs.
+
+## Receiving RPCs
+
+To update the server, we need to create a key-value service and implementation.  Recall that in gRPC, a _Server_ can handle one or more _Services_.  Again, what Protobuf would normally have generated for us we need to write ourselves.  Here is what the base service looks like:
+
+```java
+  static abstract class KeyValueServiceImplBase implements BindableService {
+    public abstract void create(
+        KvGson.CreateRequest request, StreamObserver<CreateResponse> responseObserver);
+
+    public abstract void retrieve(/*...*/);
+
+    public abstract void update(/*...*/);
+
+    public abstract void delete(/*...*/);
+
+    @Override
+    public final ServerServiceDefinition bindService() {
```

Comment:
i feel that this method might benefit from an inline comment or more discussion on what it does.

Reply:
Done.